model/release.go

```diff
@@ -11,14 +11,146 @@
 //	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 //	See the License for the specific language governing permissions and
 //	limitations under the License.
+
 package model
 
-import "time"
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	TagsMaxPerRelease = 20  // Maximum number of tags per release.
+	TagsMaxUnique     = 100 // Maximum number of unique tags.
+)
+
+var (
+	ErrTooManyTags = errors.New(
+		"the total number of tags per exceeded maximum of " +
+			strconv.Itoa(TagsMaxPerRelease),
+	)
+	ErrTooManyUniqueTags = errors.New(
+		"the total number of unique tags (" +
+			strconv.Itoa(TagsMaxUnique) +
+			") has been exceeded",
+	)
+)
+
+type Tags []Tag
+
+func (tags Tags) Validate() (err error) {
+	if len(tags) > TagsMaxPerRelease {
+		return ErrTooManyTags
+	}
+	for _, tag := range tags {
+		if err = tag.Validate(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (tags Tags) MarshalJSON() ([]byte, error) {
+	if len(tags) == 0 {
+		return []byte{'[', ']'}, nil
+	}
+	return json.Marshal([]Tag(tags))
+}
+
+func (tags *Tags) Dedup() {
+	// Deduplicate tags:
+	s := *tags
+	sort.Slice(s, func(i, j int) bool {
+		return s[i] < s[j]
+	})
+	i, j := 0, 0
+	for j < len(s) {
+		if s[i] != s[j] {
+			i++
+			s[i] = s[j]
+		}
+		j++
+	}
+	*tags = s[:i+1]
+}
```

Comment:
* I had to check that the sort works :)
* why do you have i+1 on the last line?
* where are the units? :)
alternative solution:
```suggestion
func (tags *Tags) Dedup() {
	// Deduplicate tags:
	s := *tags
	h := make(map[Tag]int)
	for _, v := range s {
		h[v] = 1
	}
	i := 0
	for k, _ := range h {
		s[i] = k
		i++
	}
	*tags = s[:i]
}
```

Reply:
> * why do you have i+1 on the last line?
`i` points to the index of the last unique element. Slicing to `i+1` also includes this last unique element in the slice. Note that `i` can be at most `len(tags)-1` when all elements are unique so it's impossible to slice out of bounds.

> * where are the units? :)
`TestReleaseTags` tests deserialization of a JSON payload with duplicate elements and therefore tests this function.

The implementation does not make any allocations and the result is always sorted.